### PR TITLE
metapath url persistence - approach 1

### PR DESCRIPTION
**Pull request by vincerubinetti**
@dongbohu I have come up with two different approaches to solve this problem. [Here is approach 2](https://github.com/greenelab/hetmech-frontend/pull/60). I am seeking feedback on which seems better. I am leaning toward approach 2, though neither solution is ideal.

@arielsvn This is what I was asking your advice for earlier today. If you have time, a brief review of the two possible approaches would be helpful. Just a gestalt review of which way seems more clean/readable/idiomatic.

(please ignore CI failure for now)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -17,6 +17,7 @@
     "react-redux": "^7.0.3",
     "react-scripts": "^3.0.1",
     "redux": "^4.0.1",
+    "redux-batched-actions": "^0.4.1",
```
Comment (vincerubinetti) [inline package.json]: This is not needed. Tried it, doesn't seem to work.

File: src/node-search.js [hunk 2]
```diff
@@ -14,7 +14,7 @@ import { MetanodeChip } from './chips.js';
 import { Tooltip } from './tooltip.js';
 import { Button } from './buttons.js';
 import { searchNodes } from './backend-query.js';
-import { updateSourceTargetNodes } from './actions.js';
+import { setState } from './actions.js';
 import { swapSourceTargetNodes } from './actions.js';
```
Comment (dongbohu) [inline src/node-search.js]: This line causes CircleCI failure. It should be removed.

File: package.json (hunk 1 above)
Comment (dongbohu) [reply package.json]: I can't really tell which one is better. @vincerubinetti, can you write down the pros and cons of each approach?

Comment (vincerubinetti) [reply package.json]: #### The inherent problem
The url must stay in sync with the relevant bits of global state. Any time the user changes the global state (searches for a new source node, checks a new metapath, swaps source/target nodes, etc), the url should update. However, when the user navigates back/forward in the history, the state must be updated from the url. In a typical React setup, components listening to changes in the global state have no way to tell whether the state change came from a user interaction or from updating the state from the url. This leads to pushing multiple redundant history entries, pushing non-sensical interim states to the history (like the source/target nodes state has updated but the metapaths hasn't yet), and may other issues.

#### Approach 1
Uses a single redux action `setState` to allow setting multiple global vars "at the same time" (without triggering multiple state and component updates). Not redux-idiomatic.

#### Approach 2
Moves the url update function to the redux reducers file, and passes a flag through actions saying whether to update the url.

Comment (dongbohu) [reply package.json]: @vincerubinetti: Thanks for the explanation. I read the code again and lean toward approach #2 too. What do you think @arielsvn?

Comment (arielsvn) [reply package.json]: Same, I think the other approach is better

Comment (vincerubinetti) [reply package.json]: I agree. This approach is terrible :octopus:

Comment (arielsvn) [reply package.json]: Still it was interesting to be able to compare them side by side, thanks for mentioning me. Enjoyed taking a peek :)

Comment (dongbohu) [reply package.json]: Thanks to both of you! @vincerubinetti: I will take a closer look at the other PR. Please feel free to close this PR.

**Pull request closed without merging**